### PR TITLE
[SwiftSyntax] Add replace method to SyntaxCollections

### PR DESCRIPTION
**Pull request by morozkin**
I started working closely with SwiftSyntax and realized that I very often face with the situation where I need to replace one element in some kind of SyntaxCollection and for doing such operation we have to call two different methods: one for delete operation and one for insert operation.

This is a typical scenario where you need to update one element in the `ArrayExprSyntax`:
```arrayExpr.elements.removing(childAt: i).inserting(newArrayElement, at: i)```

Instead of this with the new method we can rewrite it in the following way:
```arrayExpr.elements.replacing(childAt: i, with: newArrayElement)```

### CODE REVIEW

File: tools/SwiftSyntax/SyntaxCollections.swift.gyb [hunk 1]
```diff
@@ -93,6 +93,24 @@ public struct ${node.name}: _SyntaxBase {
...
+  ///   - syntax: The element to replace with.
+  ///
+  /// - Returns: A new `${node.name}` with the new element at the provided index.
+  public func replacing(childAt index: Int,
+                        with syntax: ${node.collection_element_type}) -> ${node.name} {
+    var newLayout = data.raw.layout
+    /// Make sure the index is a valid index for replacing
+    precondition((newLayout.startIndex..<newLayout.endIndex).contains(index),
+                 "replacing node at invalid index \(index)")
+    newLayout.replaceSubrange(index...index, with: [syntax])
```
Comment (harlanhaskins) [inline tools/SwiftSyntax/SyntaxCollections.swift.gyb]: Should this be `newLayout[index] = syntax.raw`?

**Review by harlanhaskins — COMMENTED** [review]
Thank you for this! If you could add a test, that’d be great!

**Comment by morozkin** [comment]
@harlanhaskins I added tests for all types of mutation methods of `SyntaxCollections`

File: test/SwiftSyntax/SyntaxCollections.swift [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import Foundation
+import StdlibUnittest
+import SwiftSyntax
+
+var SyntaxCollectionsAPI = TestSuite("SyntaxCollectionsAPI")
+
+SyntaxCollectionsAPI.test("AppendingElement") {
+    let arrayElementList = SyntaxFactory.makeArrayElementList([
+        SyntaxFactory.makeArrayElement(expression: SyntaxFactory.makeIntegerLiteral("0"), trailingComma: nil)
```
Comment (harlanhaskins) [inline test/SwiftSyntax/SyntaxCollections.swift]: `makeIntegerLiteral` returns a `TokenSyntax`, not an `ExprSyntax`. Can you rewrite this to use `makeIntegerLiteralExpr`?

You might want to make a helper:

```swift
func integerLiteralElement(_ int: Int) -> ArrayElementSyntax {
  let literal = SyntaxFactory.makeIntegerLiteral("\(int)")
  return SyntaxFactory.makeArrayElement(
    expression: SyntaxFactory.makeIntegerLiteralExpr(digits: literal),
    trailingComma: nil)
}
```

That way you always know you're constructing a full `ArrayElementSyntax`.

**Review by harlanhaskins — CHANGES_REQUESTED** [review]

**Comment by harlanhaskins** [comment]
Structurally, the tests look great!

**Review by harlanhaskins — APPROVED** [review]
LGTM. Thank you so much!

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by morozkin** [comment]
@harlanhaskins I'm sorry for the quality of my previous commits. I had errors with compilation of the swift so I didn't run and check that my code compiles and runs well. But with my latest commit I compiled the swift with my changes and checked that all tests passed successfully

**Comment by harlanhaskins** [comment]
No need to apologize 😊

I’ll re-run the tests now and we can hopefully get this merged!

@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@morozkin there's no need to feel bad. People rely on CI to find build errors all the time 😄

**Comment by harlanhaskins** [comment]
I certainly do!

**Comment by harlanhaskins** [comment]
:boat:

**Comment by harlanhaskins** [comment]
Thank you, @morozkin!

**Pull request merged**